File: Windows 8.1 app samples/FileLoggingSession Sample (Windows 8.1)/JavaScript/js/loggingScenario.js
﻿//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {

    var LoggingScenario = WinJS.Class.define(
        function () {
            Windows.UI.WebUI.WebUIApplication.addEventListener("suspending", this._onAppSuspension);
            Windows.UI.WebUI.WebUIApplication.addEventListener("resuming", this._onAppResume);
        },
        {
            doScenarioAsync: function () {
                this._setBusyStatus(true);
                try {

                    var data = { loggingScenario: this, messageIndex: 0, startFileCount: this._logFileGeneratedCount };

                    function logMessagesAndPauseAsync(data) {

                        var totalFilesCreatedSoFar = data.loggingScenario._logFileGeneratedCount - data.startFileCount;
                        if (totalFilesCreatedSoFar >= 3) {
                            // At least 3 files created, stop.
                            return new WinJS.Promise(function (completionFunction, errorFunction, progressFunction) {
                                data.loggingScenario._setBusyStatus(false);
                                completionFunction();
                            });
                        } else {

                            for (var index = 0; index < 50; index++) {

                                // Since the channel is added to the session at level Warning,
                                // the following is logged because it is logged at level LoggingLevel.Critical.
                                messageToLog =
                                    "Message=" +
                                    (++data.messageIndex) +
                                    ": Lorem ipsum dolor sit amet, consectetur adipiscing elit.In ligula nisi, vehicula nec eleifend vel, rutrum non dolor.Vestibulum ante ipsum " +
                                    "primis in faucibus orci luctus et ultrices posuere cubilia Curae; Curabitur elementum scelerisque accumsan. In hac habitasse platea dictumst.";
                                data.loggingScenario.logMessage(messageToLog, Windows.Foundation.Diagnostics.LoggingLevel.critical);

                                // Since the channel is added to the session at level Warning,
                                // the following is *not* logged because it is logged at LoggingLevel.Information.
                                messageToLog =
                                    "Message=" +
                                    (++data.messageIndex) +
                                    ": Lorem ipsum dolor sit amet, consectetur adipiscing elit.In ligula nisi, vehicula nec eleifend vel, rutrum non dolor.Vestibulum ante ipsum " +
                                    "primis in faucibus orci luctus et ultrices posuere cubilia Curae; Curabitur elementum scelerisque accumsan. In hac habitasse platea dictumst.";
                                data.loggingScenario.logMessage(messageToLog, Windows.Foundation.Diagnostics.LoggingLevel.information);

                                var value = 1000000; // one million, 7 digits, 4-bytes as an int, 14 bytes as a wide character string.
                                data.loggingScenario.logMessage("Value #" + (++data.messageIndex) + "  " + value, Windows.Foundation.Diagnostics.LoggingLevel.critical); // value is logged as 14 byte wide character string.
                                data.loggingScenario.logValuePair("Value #" + (++data.messageIndex), value,  Windows.Foundation.Diagnostics.LoggingLevel.critical); // value is logged as a 4-byte integer.
                            }

                            // Pause every once in a while to simulate application 
                            // activity outside of logging.
                            return WinJS.Promise.timeout(25).then(function () {
                                return logMessagesAndPauseAsync(data);
                            });
                        }
                    }

                    // Log messages until at least 3 log files are created. 
                    return logMessagesAndPauseAsync(data);

                } catch (e) {
                    this._setBusyStatus(false);
                    throw e;
                }
            },
            toggleLoggingEnabledDisabledAsync: function () {
                this._setBusyStatus(true);
                try {
                    var loggingScenario = this;
                    var sessionToClose = this._session
                    this._session = null;
                    if (sessionToClose !== null) {
                        return this._saveFinalLogFileAsync(sessionToClose).then(function (finalLogFIlePath) {
                            loggingScenario._dispatchStatusChanged({ type: "LogFileGeneratedAtDisable", logFilePath: finalLogFIlePath });
                            Windows.Storage.ApplicationData.current.localSettings.values["LoggingEnabled"] = false;
                            loggingScenario._dispatchStatusChanged({ type: "LoggingEnabledDisabled", enabled: false });
                            loggingScenario._setBusyStatus(false);
                            return false;
                        });
                    } else {
                        this._startLogging();
                        Windows.Storage.ApplicationData.current.localSettings.values["LoggingEnabled"] = true;
                        loggingScenario._dispatchStatusChanged({ type: "LoggingEnabledDisabled", enabled: true });
                        loggingScenario._setBusyStatus(false);
                        return new WinJS.Promise(function (completionFunction, errorFunction, progressFunction) {
                            completionFunction(true);
                        });
                    }
                } catch (e) {
                    this._setBusyStatus(false);
                    throw e;
                }
            },
            logMessage: function (message, level) {

                if (this.isLoggingChannelEnabledForLevel(level)) {

                    this._channel.logMessage(message, level);

                    // Total bytes is roughly 2 bytes per wide character.
                    // This non-critical calculation is for sample UI feedback purposes.
                    this._logMessageApproximateByteCount += (message.length * 2);
                    this._incrementMessageCount();
                }
            },
            logValuePair: function (value1, value2, level) {

                if (this.isLoggingChannelEnabledForLevel(level)) {

                    this._channel.logValuePair(value1, value2, level);

                    // Total bytes is roughly 2 bytes per wide character, plus 4 bytes for the integer.
                    // This non-critical calculation is for sample UI feedback purposes.
                    this._logMessageApproximateByteCount += (value1.length * 2) + 4; 
                    this._incrementMessageCount();
                }
            },
            isLoggingChannelEnabledForLevel: function (levelToTest) {

                if (this._channel === null) {
                    // No channel, so "false," logging is not enabled for the specified level.
                    return false;
                }

                if (this._isChannelEnabled && levelToTest >= this._channelLoggingLevel) {
                    // The channel is enabled, and the caller's level is equal to or higher
                    // than the aggregate level of all the channels' listeners.
                    return true;
                }

                // No sessions consuming events from this channel, so 'false' for "not enabled."
                return false;
            },
            isLoggingEnabled: {
                get: function () {
                    return this._session !== null;
                }
            },
            _session: null,
            _channel: null,
            _isChannelEnabled: false,
            _channelLoggingLevel: Windows.Foundation.Diagnostics.LoggingLevel.verbose,
            _logFileGeneratedCount: 0,
            _logMessageCount: 0,
            getLogMessageCount: function () {
                return this._logMessageCount;
            },
            getLogMessageApproximateByteCount: function () {
                return this._logMessageApproximateByteCount;
            },
            _incrementMessageCount: function () {

                this._logMessageCount++;

                // Update any listening UI every 500 messages.
                if (this._logMessageCount % 500 === 0) {
                    this._dispatchStatusChanged({ type: "LogMessageCountUpdate" });
                }

                return this._logMessageCount;
            },
            _logMessageApproximateByteCount: 0,
            _isSuspending: false,
            _isBusy: false,
            _setBusyStatus: function (busy) {
                if (this._isBusy === busy) {
                    return;
                }
                this._isBusy = busy;
                if (busy) {
                    this._dispatchStatusChanged({ type: "BusyStatusChanged" });
                }
            },
            getBusyStatus: function () {
                return this._isBusy;
            },            
            _onAppSuspension: function (suspendingEventArgs) {

                var loggingScenario = LoggingScenario.loggingScenarioSingleton;

                // Get a deferral before performing any async operations
                // to avoid suspension prior to LoggingScenario completing 
                // PrepareToSuspendAsync().
                var deferral = suspendingEventArgs.suspendingOperation.getDeferral();
                loggingScenario._prepareToSuspendAsync().then(function () {
                    // From LoggingScenario's perspective, it's now okay to 
                    // suspend, so release the deferral. 
                    deferral.complete();
                });
            },
            _onAppResume: function (resmingEventArgs) {
                var loggingScenario = LoggingScenario.loggingScenarioSingleton;
                loggingScenario.resumeLoggingIfApplicable();
            },
            _startLogging: function () {

                // _startLogging called in two places: resume and when logging is 
                // enabled. It is assumed both cannot occur at the same time so no
                // interlocked or synchronization is used when checking 'session' here. 
                if (this._session === null)
                {
                    this._session = new Windows.Foundation.Diagnostics.FileLoggingSession("AppSession1");
                    this._session.addEventListener("logfilegenerated", this._onLogFileGenerated);
                }

                if (this._channel === null)
                {
                    this._channel = new Windows.Foundation.Diagnostics.LoggingChannel("AppChannel1");
                    this._channel.addEventListener("loggingenabled", this._onLoggingEnabled);
                }

                // This sample adds the channel at level "warning" to 
                // demonstrated how messages logged at more verbose levels
                // are ignored by the session. 
                this._session.addLoggingChannel(this._channel, Windows.Foundation.Diagnostics.LoggingLevel.warning);
            },
            _onLoggingEnabled: function (args) {

                var loggingScenario = LoggingScenario.loggingScenarioSingleton;

                // Given the sessions consuming the channel, 
                // the channel's enablement status or level
                // has changed. Save those values.
                loggingScenario._isChannelEnabled = args.target.enabled;
                loggingScenario._channelLoggingLevel = args.target.level;
            },
            _onLogFileGenerated: function (args) {

                var loggingScenario = LoggingScenario.loggingScenarioSingleton;

                loggingScenario._logFileGeneratedCount++;
                loggingScenario._moveFileToRepositoryAsync(args.file).then(function (finalLogFileFullPath) {
                    if (loggingScenario._isSuspending === false) {
                        loggingScenario._dispatchStatusChanged({ type: "LogFileGenerated", logFilePath: finalLogFileFullPath });
                    }
                });
            },
            resumeLoggingIfApplicable: function () {

                // If logging is already enabled, nothing to do.
                if (this.isLoggingEnabled) {
                    return true;
                }

                var loggingEnabled;
                if (Windows.Storage.ApplicationData.current.localSettings.values.hasKey("LoggingEnabled")) {
                    loggingEnabled = Windows.Storage.ApplicationData.current.localSettings.values.hasKey("LoggingEnabled");
                } else {
                    Windows.Storage.ApplicationData.current.localSettings.values["LoggingEnabled"] = true;
                    loggingEnabled = true;
                }

                if (loggingEnabled) {
                    this._startLogging();
                }
                
                if (Windows.Storage.ApplicationData.current.localSettings.values.hasKey("LogFileGeneratedBeforeSuspend")) {

                    var logFileGeneratedBeforeSuspend = Windows.Storage.ApplicationData.current.localSettings.values["LogFileGeneratedBeforeSuspend"];
                    if (logFileGeneratedBeforeSuspend !== null) {
                        this._dispatchStatusChanged( { type: "LogFileGeneratedAtSuspend", logFilePath: logFileGeneratedBeforeSuspend });
                    }
                    Windows.Storage.ApplicationData.current.localSettings.values["LogFileGeneratedBeforeSuspend"] = null;
                }
            },
            _getLogRepositoryFolderAsync: function () {
                return Windows.Storage.ApplicationData.current.localFolder.createFolderAsync("OurAppLogFiles", Windows.Storage.CreationCollisionOption.openIfExists);
            },
            _getAllLogRepositoryFileNamesAsync: function () {
                return this._getLogRepositoryFolderAsync().then(function (logRepositoryFolder) {
                    var queryOptions = new Windows.Storage.Search.QueryOptions(Windows.Storage.Search.CommonFileQuery.orderBySearchRank, null);
                    queryOptions.applicationSearchFilter = "System.FileName:~\"OurAppLog-????.etl\"";
                    var queryResult = logRepositoryFolder.createFileQueryWithOptions(queryOptions);
                    return queryResult.getFilesAsync();
                });
            },
            _getNextLogFileNumberAsync: function () {

                // Get all files in this sample's log repository, and 
                // find the file with the highest number. Return
                // that number plus 1 as the next file number.
                return this._getAllLogRepositoryFileNamesAsync().then(function (files) {
                    var extractNumberRegExPattern = new RegExp("OurAppLog-(\\d{4})\\.etl", "i");
                    var lastNumber = 0;
                    files.forEach(function (file) {
                        var matches = extractNumberRegExPattern.exec(file.name);
                        if (matches !== null) {
                            var fileNumber = parseInt(matches[1], 10);
                            if (isNaN(fileNumber) == false && fileNumber > lastNumber) {
                                lastNumber = fileNumber;
                            }
                        }
                    });

                    // return the next available log file number.
                    return lastNumber + 1;
                });
            },
            _getNextLogFileNameAsync: function () {
                return this._getNextLogFileNumberAsync().then(function (nextLogFileNumber) {
                    return "OurAppLog-" + ("000" + nextLogFileNumber).substr(-4) + ".etl";
                });
            },
            _moveFileToRepositoryAsync: function (sourceFile) {
                var data = { loggingScenario: this, logRepositoryFolder: null, nextLogFileName: null };
                return this._getLogRepositoryFolderAsync().then(function (logRepositoryFolder) {
                    data.logRepositoryFolder = logRepositoryFolder;
                    return data.loggingScenario._getNextLogFileNameAsync();
                }).then(function (nextLogFileName) {
                    data.nextLogFileName = nextLogFileName;
                    return sourceFile.moveAsync(data.logRepositoryFolder, nextLogFileName, Windows.Storage.NameCollisionOption.failIfExists);
                }).then(function () {
                    return data.logRepositoryFolder.path + "\\" + data.nextLogFileName;
                });
            },
            _saveFinalLogFileAsync: function (sessionToCloseAndSave) {
                /// <summary>Call the CloseAndSaveToFileAsync() method of a FileLoggingSession, and move the log file to this app's log file repository.</summary>
                /// <param name="sessionToCloseAndSave" type="Windows.Foundation.Diagnostics.FileLoggingSession">The FileLoggingSession.</param>
                /// <returns type="WinJS.Promise">A Promise to return the final path of the log file, or null if there was no final log file.</returns>
                
                var data = { loggingScenario: this };
                return sessionToCloseAndSave.closeAndSaveToFileAsync().then(function (finalLogOfSession) {

                    // FileLoggingSession.closeAndSaveToFileAsync can return null if nothing
                    // has been logged to the session since the last log file was saved.
                    if (finalLogOfSession !== null) {

                        // There was a log file, so move it to the app's repository.
                        return data.loggingScenario._moveFileToRepositoryAsync(finalLogOfSession);
                    }

                    // There was no log file, so simply return the 'null' path to indicate there was no log file.
                    return new WinJS.Promise(function (completionFunction, errorFunction, progressFunction) {
                        completionFunction(finalLogOfSession); // finalLogOfSession == null in this case.
                    });
                }).then(function (finalAppLogFileFullPath) {

                    // return the log file path or null if there was none.
                    return finalAppLogFileFullPath;
                });
            },
            _prepareToSuspendAsync: function () {
                /// <summary>Save any final log file and save app state before suspending.</summary>
                /// <returns type="WinJS.Promise">A Promise.</returns>
                var sessionToClose = this._session
                this._session = null;
                if (sessionToClose !== null) {
                    this._isSuspending = true;
                    var data = { loggingScenario: this };
                    return this._saveFinalLogFileAsync(sessionToClose).then(function (finalLogFIlePath) {
                        Windows.Storage.ApplicationData.current.localSettings.values["LoggingEnabled"] = true;
                        Windows.Storage.ApplicationData.current.localSettings.values["LogFileGeneratedBeforeSuspend"] = finalLogFIlePath;
                    }).then(function () {
                        data.loggingScenario._isSuspending = false;
                    });
                } else {
                    Windows.Storage.ApplicationData.current.localSettings.values["LoggingEnabled"] = false;
                    Windows.Storage.ApplicationData.current.localSettings.values["LogFileGeneratedBeforeSuspend"] = null;                    
                    return new WinJS.Promise(function (completionFunction, errorFunction, progressFunction) {
                        completionFunction(); 
                    });
                }
            },
            _dispatchStatusChanged: function (statusArgs) {
                this.dispatchEvent("statusChanged", statusArgs);
            }
        },
        {
            // LoggingScenario static members:

            _loggingScenarioSingleton: null,
            loggingScenarioSingleton: {
                get: function () {
                    if (!this._loggingScenarioSingleton) {
                        this._loggingScenarioSingleton = new LoggingScenario()
                    }
                    return this._loggingScenarioSingleton;
                }
            }
        });
    
    WinJS.Class.mix(LoggingScenario, WinJS.Utilities.eventMixin);

    // Export public methods & controls
    WinJS.Namespace.define("LoggingScenario", {
        LoggingScenario: LoggingScenario
    });

})();

File: Windows 8.1 app samples/FileLoggingSession Sample (Windows 8.1)/C++/LoggingScenario.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
//
//*********************************************************

//
// LoggingScenario.cpp
// Implementation of the LoggingScenario class
//

#include "pch.h"
#include "LoggingScenario.h"

using namespace concurrency;
using namespace Windows::Storage;
using namespace Windows::Storage::Search;
using namespace Platform::Collections;
using namespace SDKSample;
using namespace SDKSample::FileLoggingSession;

LoggingScenario^ LoggingScenario::loggingScenario;


LoggingScenario::LoggingScenario()
{
	_logFileGeneratedCount = 0;
	_logMessageCount = 0;
    _logMessageApproximateByteCount = 0;
    _isSuspending = false;
    _isBusy = false;

    SDKSample::App::Current->Suspending += ref new Windows::UI::Xaml::SuspendingEventHandler(this, &LoggingScenario::OnAppSuspending);
    SDKSample::App::Current->Resuming += ref new Windows::Foundation::EventHandler<Platform::Object ^>(this, &LoggingScenario::OnAppResuming);

    // If the app is being launched (not resumed), the 
    // following call will activate logging if it had been 
    // activated during the last suspend. 
    ResumeLoggingIfApplicable();
}

LoggingScenario::~LoggingScenario()
{

}

void LoggingScenario::OnAppSuspending(Platform::Object ^sender, Windows::ApplicationModel::SuspendingEventArgs ^e)
{
    (void) sender;    // Unused parameter

    // Get a deferral before performing any async operations
    // to avoid suspension prior to LoggingScenario completing 
    // PrepareToSuspendAsync().
    auto deferral = e->SuspendingOperation->GetDeferral();
    PrepareToSuspendAsync().then([=](task<void> previousTask)
    {
        // From LoggingScenario's perspective, it's now okay to 
        // suspend, so release the deferral. 
        deferral->Complete();
    });
}

void LoggingScenario::OnAppResuming(Platform::Object ^sender, Platform::Object ^args)
{
    // If logging was active at the last suspend,
    // ResumeLoggingIfApplicable will re-activate 
    // logging.
    ResumeLoggingIfApplicable();
}

void LoggingScenario::OnChannelLoggingEnabled(ILoggingChannel^ sender, Platform::Object^ args)
{
	_isChannelEnabled = sender->Enabled;
	_channelLoggingLevel = sender->Level;
}

void LoggingScenario::OnLogFileGenerated(IFileLoggingSession^ sender, LogFileGeneratedEventArgs^ args)
{
    InterlockedIncrement(&_logFileGeneratedCount);
    create_task(MoveFileToRepositoryAsync(args->File)).then([this](Platform::String^ finalLogFileFullPath)
    {
        if (_isSuspending == false)
        {
            StatusChanged(this, ref new LoggingScenarioEventArgs(LoggingScenarioEventType::LogFileGenerated, finalLogFileFullPath));
        }
    });
}

task<StorageFolder^> LoggingScenario::GetLogRepositoryFolderAsync()
{
	return create_task(ApplicationData::Current->LocalFolder->CreateFolderAsync(LOG_FILE_REPOSITORY_FOLDER_NAME, CreationCollisionOption::OpenIfExists));
}

task<IVectorView<StorageFile^>^> LoggingScenario::GetAllLogRepositoryFileNamesAsync()
{
	return GetLogRepositoryFolderAsync().then([this](StorageFolder^ logRepositoryFolder)
	{
		QueryOptions^ queryOptions = ref new QueryOptions(CommonFileQuery::OrderBySearchRank, nullptr);
		queryOptions->ApplicationSearchFilter = "System.FileName:~\"" + LOG_FILE_BASE_FILE_NAME + "-????.etl\"";
		StorageFileQueryResult^ queryResult = logRepositoryFolder->CreateFileQueryWithOptions(queryOptions);
		return create_task(queryResult->GetFilesAsync());
	});
}

task<int> LoggingScenario::GetNextLogFileNumberAsync()
{
	return GetAllLogRepositoryFileNamesAsync().then([=](IVectorView<StorageFile^>^ files)
	{
		Platform::String^ extractNumberRegExPattern = LOG_FILE_BASE_FILE_NAME + "-(\\d{4})\\.etl";
		std::wregex extractNumberRegEx(extractNumberRegExPattern->Data());
		int lastNumber = 0;
		std::for_each(begin(files), end(files), [=,&lastNumber](StorageFile^ file)
		{
			std::wstring fileName(file->Name->Data());
			std::wsmatch m;
			if (std::regex_search(fileName, m, extractNumberRegEx))
			{
				int tempNumber = std::stoi(m[1].str());
				if (tempNumber > lastNumber)
				{
					lastNumber = tempNumber;
				}
			}
		});

		return lastNumber + 1;
	});
}

task<Platform::String^> LoggingScenario::GetNextLogFileNameAsync()
{
	return GetNextLogFileNumberAsync().then([=](int nextLogFileNumber)
	{
		std::wostringstream nextLogFileName;
        nextLogFileName 
            << std::wstring(LOG_FILE_BASE_FILE_NAME->Data())
            << L'-'
            << std::setfill(L'0')
            << std::setw(4)
            << nextLogFileNumber
            << L".etl";
        return ref new Platform::String(nextLogFileName.str().c_str());
	});
}

task<Platform::String^> LoggingScenario::MoveFileToRepositoryAsync(StorageFile^ sourceFile)
{
    struct ContinuationData
    {
        StorageFolder^ logRepositoryFolder;
        Platform::String^ nextLogFileName;
    };
    auto data = make_shared<ContinuationData>();

    return GetLogRepositoryFolderAsync().then([this, data](StorageFolder^ logRepositoryFolder)
	{
        _lockForMoveFile.lock();
        data->logRepositoryFolder = logRepositoryFolder;
        return GetNextLogFileNameAsync();
    }).then([this, sourceFile, data](Platform::String^ nextLogFileName)
    {
        data->nextLogFileName = nextLogFileName;
        return sourceFile->MoveAsync(data->logRepositoryFolder, nextLogFileName, Windows::Storage::NameCollisionOption::FailIfExists);
    }).then([this, data](task<void> previousTask)
    {
        _lockForMoveFile.unlock();
        return data->logRepositoryFolder->Path + "\\" + data->nextLogFileName;
    });
}

task<Platform::String^> LoggingScenario::SaveFinalLogFileAsync(Windows::Foundation::Diagnostics::FileLoggingSession^ sessionToCloseAndSave)
{
    return create_task(sessionToCloseAndSave->CloseAndSaveToFileAsync()).then([this](StorageFile^ finalFileBeforeSuspend)
    {
        if (finalFileBeforeSuspend != nullptr)
        {
            return create_task(MoveFileToRepositoryAsync(finalFileBeforeSuspend));
        }

        return task<Platform::String^>([]() -> Platform::String^
        {
            return nullptr;
        });     
    }).then([this](Platform::String^ finalAppLogFileFullPath)
    {
        return finalAppLogFileFullPath;
    });
}

task<bool> LoggingScenario::ToggleLoggingEnabledDisabledAsync()
{
    IsBusy = true;
    try
    {
        Windows::Foundation::Diagnostics::FileLoggingSession^ sessionCopy;
        sessionCopy = InterlockedExchangeRefValues(&_session, (Windows::Foundation::Diagnostics::FileLoggingSession^)nullptr);
        if (sessionCopy != nullptr)
        {
            return create_task(SaveFinalLogFileAsync(sessionCopy)).then([this, sessionCopy](Platform::String^ finalLogFilePath)
            {
                delete sessionCopy;
                StatusChanged(this, ref new LoggingScenarioEventArgs(LoggingScenarioEventType::LogFileGeneratedAtDisable, finalLogFilePath));
                SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, false);
                StatusChanged(this, ref new LoggingScenarioEventArgs(false));
                IsBusy = false;
                return false;
            });
        }
        else
        {
            StartLogging();
            SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, true);
            StatusChanged(this, ref new LoggingScenarioEventArgs(true));
            IsBusy = false;
            return task<bool>([]() -> bool
            {
                return true;
            });
        }
    }
    catch (...)
    {
        IsBusy = false;
        throw;
    }
}

void LoggingScenario::StartLogging()
{
    // StartLogging called in two places: resume and when logging is 
    // enabled. It is assumed both cannot occur at the same time so no
    // interlocked or synchronization is used when checking 'session' here. 
    if (_session == nullptr)
    {
        _session = ref new Windows::Foundation::Diagnostics::FileLoggingSession(DEFAULT_SESSION_NAME);
        _session->LogFileGenerated += ref new Windows::Foundation::TypedEventHandler<IFileLoggingSession ^, LogFileGeneratedEventArgs ^>(this, &LoggingScenario::OnLogFileGenerated);
    }

    if (_channel == nullptr)
    {
        _channel = ref new Windows::Foundation::Diagnostics::LoggingChannel(DEFAULT_CHANNEL_NAME);
        _channel->LoggingEnabled += ref new Windows::Foundation::TypedEventHandler<Windows::Foundation::Diagnostics::ILoggingChannel ^, Platform::Object ^>(this, &LoggingScenario::OnChannelLoggingEnabled);
    }

    // This sample adds the channel at level "warning" to 
    // demonstrated how messages logged at more verbose levels
    // are ignored by the session. 
    _session->AddLoggingChannel(_channel, LoggingLevel::Warning);
}

void LoggingScenario::LogMessage(Platform::String^ message, Windows::Foundation::Diagnostics::LoggingLevel level)
{
    if (IsLoggingChannelEnabledForLevel(level))
    {
        _channel->LogMessage(message, level);

        // Total bytes is roughly 2 bytes per wide character.
        // This non-critical calculation is for sample UI feedback purposes.
        LogMessageApproximateByteCount += message->Length() * 2;
        LogMessageCount++;
    }
}

void LoggingScenario::LogValuePair(Platform::String^ value1, int value2, Windows::Foundation::Diagnostics::LoggingLevel level)
{
    if (IsLoggingChannelEnabledForLevel(level))
    {
        _channel->LogValuePair(value1, value2, level);

        // Total bytes is roughly 2 bytes per wide character, plus 4 bytes for the integer.
        // This non-critical calculation is for sample UI feedback purposes.
        LogMessageApproximateByteCount += (value1->Length() * 2) + 4; 
        LogMessageCount++;
    }
}


bool LoggingScenario::IsLoggingChannelEnabledForLevel(Windows::Foundation::Diagnostics::LoggingLevel level)
{
    if (_channel == nullptr)
    {
        //
        // No channel, so "false" logging is not enabled for the specified level.
        //

        return false;
    }

    if (_isChannelEnabled && level >= _channelLoggingLevel)
    {
        //
        // The channel is enabled, and the caller's level is equal to or higher
        // than the aggregate level of all the channels' listeners.
        //

        return true;
    }

    //
    // No sessions consuming events from this channel, so 'false' for "not enabled."
    //

    return false;
}

task<void> LoggingScenario::PrepareToSuspendAsync()
{
    Windows::Foundation::Diagnostics::FileLoggingSession^ sessionCopy;
    sessionCopy = InterlockedExchangeRefValues(&_session, (Windows::Foundation::Diagnostics::FileLoggingSession^)nullptr);
    if (sessionCopy != nullptr)
    {
        _isSuspending = true;
        return create_task(SaveFinalLogFileAsync(sessionCopy)).then([this, sessionCopy](Platform::String^ finalLogFilePath)
        {
            delete sessionCopy;
            SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, true);
            SetAppLocalSettingsValue(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME, finalLogFilePath);          
        }).then([this](task<void> previousTask)
        {
            _isSuspending = false;
        });
    }
    else
    {
        SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, false);
        SetAppLocalSettingsValue(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME, nullptr);
        return task<void>([]() -> void {});
    }
}

void LoggingScenario::ResumeLoggingIfApplicable()
{
    bool loggingEnabled;
    if (IsAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME))
    {
        loggingEnabled = GetAppLocalSettingsValueAsBool(LOGGING_ENABLED_SETTING_KEY_NAME);
    }
    else
    {         
        SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, true);
        loggingEnabled = true;
    }

    if (loggingEnabled)
    {
        StartLogging();
    }

    // When the sample suspends, it retains state as to whether or not it had
    // generated a new log file at the last suspension. This allows any
    // UI to be updated on resume to reflect that fact. 
    if (IsAppLocalSettingsValue(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME))
    {
        Platform::String^ logFileGeneratedBeforeSuspend = GetAppLocalSettingsValueAsString(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME);
        if (logFileGeneratedBeforeSuspend != nullptr)
        {
            StatusChanged(this, ref new LoggingScenarioEventArgs(LoggingScenarioEventType::LogFileGeneratedAtSuspend, logFileGeneratedBeforeSuspend));
        }
        SetAppLocalSettingsValue(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME, nullptr);
    }
}

task<void> LoggingScenario::DoScenarioAsync()
{
    IsBusy = true;
    try
    {
        auto workItemDelegate = [this](Windows::Foundation::IAsyncAction^ workItem)
        {
            const int NUMBER_OF_LOG_FILES_TO_GENERATE = 3;
            int messageIndex = 0;
            int startFileCount = LogFileGeneratedCount;

            while (LogFileGeneratedCount - startFileCount < NUMBER_OF_LOG_FILES_TO_GENERATE)
            {
                // Since the channel is added to the session at level Warning,
                // the following is logged because it is logged at level LoggingLevel.Critical.
                Platform::String^ messageToLog =
                    "Message=" +
                    (++messageIndex).ToString() +
                    ": Lorem ipsum dolor sit amet, consectetur adipiscing elit.In ligula nisi, vehicula nec eleifend vel, rutrum non dolor.Vestibulum ante ipsum " +
                    "primis in faucibus orci luctus et ultrices posuere cubilia Curae; Curabitur elementum scelerisque accumsan. In hac habitasse platea dictumst.";
                LogMessage(messageToLog, LoggingLevel::Critical);

                // Since the channel is added to the session at level Warning,
                // the following is *not* logged because it is logged at LoggingLevel.Information.
                messageToLog = 
                    "Message=" +
                    (++messageIndex).ToString() +
                    ": Lorem ipsum dolor sit amet, consectetur adipiscing elit.In ligula nisi, vehicula nec eleifend vel, rutrum non dolor.Vestibulum ante ipsum " +
                    "primis in faucibus orci luctus et ultrices posuere cubilia Curae; Curabitur elementum scelerisque accumsan. In hac habitasse platea dictumst.";
                LogMessage(messageToLog, LoggingLevel::Information);

                int value = 1000000; // one million, 7 digits, 4-bytes as an int, 14 bytes as a wide character string.
                LogMessage("Value #" + (++messageIndex).ToString() + "  " + value.ToString(), LoggingLevel::Critical); // value is logged as 14 byte wide character string.
                LogValuePair("Value #" + (++messageIndex).ToString(), value, LoggingLevel::Critical); // value is logged as a 4-byte integer.

                //
                // Pause every once in a while to simulate application 
                // activity outside of logging.
                //

                if (messageIndex % 50 == 0)
                {                   
                    complete_after(5).wait();
                }
            }

            IsBusy = false;
        };

        auto workItemHandler = ref new Windows::System::Threading::WorkItemHandler(workItemDelegate);
        return create_task(Windows::System::Threading::ThreadPool::RunAsync(workItemHandler, Windows::System::Threading::WorkItemPriority::Normal)).then([](task<void> previousTask){

            try
            {
                previousTask.get();
            }
            catch (...)
            {
                OutputDebugString(L"Exception during DoScenario.");
                throw;
            }
        });
    }
    catch (...)
    {
        IsBusy = false;
        throw;
    }
}
